docs/src/start/logging-packets.md

```diff
@@ -0,0 +1,295 @@
+    pub ipv4_address: u32,
+    pub action: u32,
+}
+
+#[cfg(feature = "user")]
+unsafe impl aya::Pod for PacketLog {}
+```
+
+> 💡 **HINT: Struct Alignment**
+>
+> Structs must be aligned. You can either do this manually by ensuring you use `u32` fields, or some multiple of a `u32` (2 x `u16`, 4x `u8`, etc...). Alternatively you may use `#[repr(packed)]`. If you do not do this, the eBPF verifier will get upset and emit an `invalid indirect read from stack` error.
+
+We also need to ensure that the `aya::Pod` trait has been implemented for our data structure so we can use it in a typed map on the user-space side.
+
+### eBPF: Map Creation
+
+Let's create a map called `EVENTS` in `myapp-ebpf/src/main.rs`
+
+```rust,ignore
+use aya_bpf::macros::map;
+use aya_bpf::maps::PerfMap;
+use myapp-common::PacketLog;
+
+#[map(name = "EVENTS")]
+static mut EVENTS: PerfMap<PacketLog> = PerfMap::<PacketLog>::with_max_entries(1024, 0);
+```
+
+When the eBPF program is loaded by Aya, the map will be created for us.
+
+### Userspace: Map Creation
+
+After our call to `probe.attach()` we'll add the following code.
+
+```rust,ignore
+use aya::maps::AsyncPerfEventArray;
+
+let mut perf_array = AsyncPerfEventArray::try_from(bpf.map_mut("EVENTS")?)?;
+```
+
+Our `perf_array` is a mutable reference to the map that was created after the XDP program was loaded by Aya.
+
+## Writing Data
+
+Now we've got our maps set up, let's add some data!
+
+### Generating Bindings To vmlinux.h
+
+To get useful data to add to our maps, we first need some useful data structures to populate with data from the `XdpContext`.
+We want to log the Source IP Address of incoming traffic, so we'll need to:
+
+1. Read the Ethernet Header to determine if this is an IPv4 Packet
+1. Read the Source IP Address from the IPv4 Header
+
+The two structs in the kernel for this are `ethhdr` from `uapi/linux/if_ether.h` and `iphdr` from `uapi/linux/ip.h`.
+If I were to use bindgen to generate Rust bindings for those headers, I'd be tied to the kernel version of the system that I'm developing on.
+This is where `aya-gen` comes in to play. It can easily generate bindings for using `/sys/kernel/btf/vmlinux`.
+What's special about these bindings is that we are able to benefit from "Compile-Once, Run-Everywhere" support.
+While it's unlikely that these structs will change because they are part of uapi and one of the golden rules of kernel development is to "never break user-space", it's not impossible that this may happen.
+CO-RE wraps access to struct fields to ensure that, even if the field offset moves, you'll always be able to access it by name.
+
+Lets use `myapp-ebpf/build.rs` to generate these bindings for us:
+
+```rust,ignore
+use std::env;
+use std::io::Write;
+use std::fs::File;
+use std::path::{Path, PathBuf};
+use aya_gen::btf_types;
+
+fn main() {
+    // Tell Cargo that if the given file changes, to rerun this build script.
+    println!("cargo:rerun-if-changed=/sys/kernel/btf/vmlinux");
+
```

Comment:
I never thought of this, and it's kind of cool 😂 But if you were using BTF, this wouldn't be needed. With BTF you could use `aya-gen` from the command line, commit the output, and then let aya figure out the layout differences at runtime using BTF.

Reply:
You're right. The only downside to running `aya-gen` on the command-line is that the configuration never gets checked in to source control, so you (or your shell history, or a README) have to remember what the full list is.

I like `build.rs` but the constraint of `you shall not put files anywhere but OUT_DIR` is a bit annoying for this use case.

As we already have `xtask`, it might be worth doing this via `xtask` instead. Then everything ends up in source control.

Reply:
Good point, yes it should go in xtask

Reply:
fixed.